persistence/pre_persistence/persistence.go

```diff
@@ -0,0 +1,306 @@
+	"github.com/syndtr/goleveldb/leveldb/comparer"
+	"github.com/syndtr/goleveldb/leveldb/memdb"
+	"github.com/syndtr/goleveldb/leveldb/util"
+	"google.golang.org/protobuf/proto"
+)
+
+var (
+	FirstSavePointKey                                        = []byte("first_savepoint_key")
+	DeletedPrefixKey                                         = []byte("deleted/")
+	BlockPrefix                                              = []byte("block/")
+	TransactionKeyPrefix                                     = []byte("transaction/")
+	PoolPrefixKey                                            = []byte("pool/")
+	AccountPrefixKey                                         = []byte("account/")
+	AppPrefixKey                                             = []byte("app/")
+	UnstakingAppPrefixKey                                    = []byte("unstaking_app/")
+	ServiceNodePrefixKey                                     = []byte("service_node/")
+	UnstakingServiceNodePrefixKey                            = []byte("unstaking_service_node/")
+	FishermanPrefixKey                                       = []byte("fisherman/")
+	UnstakingFishermanPrefixKey                              = []byte("unstaking_fisherman/")
+	ValidatorPrefixKey                                       = []byte("validator/")
+	UnstakingValidatorPrefixKey                              = []byte("unstaking_validator/")
+	ParamsPrefixKey                                          = []byte("params/")
+	_                             modules.PersistenceModule  = &PrePersistenceModule{}
+	_                             modules.PersistenceContext = &PrePersistenceContext{}
+)
+
+type PrePersistenceModule struct {
+	CommitDB *memdb.DB
+	Mempool  types.Mempool
+	//pocketBusMod modules.Bus  TODO: add back in once bus is integrated
+	Cfg *config.Config
+}
+
+func NewPrePersistenceModule(commitDB *memdb.DB, mempool types.Mempool, cfg *config.Config) *PrePersistenceModule {
+	return &PrePersistenceModule{CommitDB: commitDB, Mempool: mempool, Cfg: cfg}
+}
+
+func (m *PrePersistenceModule) NewContext(height int64) (modules.PersistenceContext, error) {
+	newDB := NewMemDB()
+	it := m.CommitDB.NewIterator(&util.Range{Start: HeightKey(height, nil), Limit: HeightKey(height+1, nil)})
+	it.First()
+	defer it.Release()
+	for ; it.Valid(); it.Next() {
+		err := newDB.Put(KeyFromHeightKey(it.Key()), it.Value())
+		if err != nil {
+			return nil, err
+		}
+	}
+	context := &PrePersistenceContext{
+		Height: height,
+		Parent: m,
+		DBs:    make([]*memdb.DB, 0),
+	}
+	context.DBs = append(context.DBs, newDB)
+	return context, nil
+}
+
+func (m *PrePersistenceModule) GetCommitDB() *memdb.DB {
+	return m.CommitDB
+}
+
+type PrePersistenceContext struct {
+	Height     int64
+	Parent     modules.PersistenceModule
+	SavePoints map[string]int // TODO save points not entirely implemented. Happy path only for now, rollbacks for later
+	DBs        []*memdb.DB
+}
+
+func (m *PrePersistenceContext) GetLatestBlockHeight() (uint64, error) {
```

Comment:
_Educational question_: why does the height have to be a `uint64`

Reply:
As opposed to what? A string? Height is a commonly used thing and we want to strike a balance between space saving (for blockchain storage) & future proofing